### PR TITLE
Add Kerberos stats

### LINKED ISSUE #68: /proc/fs/lustre/sptlrpc/encrypt_page_pools stats
There's another missing stats file used for Lustre's GSS/Kerberos support:
```
# cat /proc/fs/lustre/sptlrpc/encrypt_page_pools
physical pages:          16465072
pages per pool:          512
max pages:               2058134
max pools:               4020
total pages:             0
total free:              0
idle index:              100/100
last shrink:             1015775s
last access:             1015775s
max pages reached:       0
grows:                   0
grows failure:           0
shrinks:                 0
cache access:            0
cache missing:           0
low free mark:           0
max waitqueue depth:     0
max wait time:           0/1000
out of mem:             0
```
The implementation is in `lustre-release.git/lustre/ptlrpc/sec_bulk.c`:
```
int sptlrpc_proc_enc_pool_seq_show(struct seq_file *m, void *v)
{
[...]
        seq_printf(m, "physical pages:          %lu\n"
                   "pages per pool:          %lu\n"
                   "max pages:               %lu\n"
                   "max pools:               %u\n"
                   "total pages:             %lu\n"
                   "total free:              %lu\n"
                   "idle index:              %lu/100\n"
                   "last shrink:             %lds\n"
                   "last access:             %lds\n"
                   "max pages reached:       %lu\n"
                   "grows:                   %u\n"
                   "grows failure:           %u\n"
                   "shrinks:                 %u\n"
                   "cache access:            %lu\n"
                   "cache missing:           %lu\n"
                   "low free mark:           %lu\n"
                   "max waitqueue depth:     %u\n"
                   "max wait time:           "CFS_TIME_T"/%lu\n"
                   "out of mem:             %lu\n",
                   totalram_pages, PAGES_PER_POOL,
                   page_pools.epp_max_pages,
                   page_pools.epp_max_pools,
                   page_pools.epp_total_pages,
                   page_pools.epp_free_pages,
                   page_pools.epp_idle_idx,
                   (long)(ktime_get_seconds() - page_pools.epp_last_shrink),
                   (long)(ktime_get_seconds() - page_pools.epp_last_access),
                   page_pools.epp_st_max_pages,
                   page_pools.epp_st_grows,
                   page_pools.epp_st_grow_fails,
                   page_pools.epp_st_shrinks,
                   page_pools.epp_st_access,
                   page_pools.epp_st_missings,
                   page_pools.epp_st_lowfree,
                   page_pools.epp_st_max_wqlen,
                   page_pools.epp_st_max_wait,
                   msecs_to_jiffies(MSEC_PER_SEC),
                   page_pools.epp_st_outofmem);
[...]
}
```

### PR DESCRIPTION
**Pull request by roclark**
Fixes #68

Signed-Off-By: Robert Clark <robert.d.clark@hpe.com>

### CODE REVIEW

**Comment by roclark** [comment]
`lustre_exporter_test.go` includes the new metrics that are generated. As usual, really looking to see if the metric names, help text, and types are correct.

File: sources/procfs.go [hunk 1]
```diff
@@ -161,6 +180,24 @@ func (s *lustreProcfsSource) generateOSTMetricTemplates() {
 			{"pool/shrink_request", "shrink_requests_total", "Number of shrinks that have been requested", s.counterMetric, false},
 			{"pool/slv", "server_lock_volume", "Current value for server lock volume (SLV)", s.gaugeMetric, false},
 		},
+		"sptlrpc": {
+			{"encrypt_page_pools", "physical_pages_total", physicalPagesHelp, s.gaugeMetric, false},
```
Comment (joehandzik) [inline sources/procfs.go]: This shouldn't be _total if it's a gauge, right?

File: sources/procfs.go [hunk 2]
```diff
@@ -161,6 +180,24 @@ func (s *lustreProcfsSource) generateOSTMetricTemplates() {
...
+			{"encrypt_page_pools", "pages_per_pool_total", pagesPerPoolHelp, s.gaugeMetric, false},
```
Comment (joehandzik) [inline sources/procfs.go]: Also should not be _total.

File: sources/procfs.go [hunk 3]
```diff
@@ -161,6 +180,24 @@ func (s *lustreProcfsSource) generateOSTMetricTemplates() {
...
+			{"encrypt_page_pools", "maximum_pages", maxPagesHelp, s.gaugeMetric, false},
+			{"encrypt_page_pools", "maximum_pools", maxPoolsHelp, s.gaugeMetric, false},
+			{"encrypt_page_pools", "pages_currently_in_pools", totalPagesHelp, s.gaugeMetric, false},
```
Comment (joehandzik) [inline sources/procfs.go]: The currently bit seems excessive to me. Does 'pages_in_pools' work for you?

File: sources/procfs.go [hunk 4]
```diff
@@ -161,6 +180,24 @@ func (s *lustreProcfsSource) generateOSTMetricTemplates() {
...
+			{"encrypt_page_pools", "free_pages", totalFreeHelp, s.gaugeMetric, false},
+			{"encrypt_page_pools", "maximum_pages_reached_total", maxPagesReachedHelp, s.counterMetric, false},
+			{"encrypt_page_pools", "grows_total", growsHelp, s.counterMetric, false},
+			{"encrypt_page_pools", "grows_failure_total", growsFailureHelp, s.counterMetric, false},
+			{"encrypt_page_pools", "shrinks_total", shrinksHelp, s.counterMetric, false},
+			{"encrypt_page_pools", "cache_access_total", cacheAccessHelp, s.counterMetric, false},
+			{"encrypt_page_pools", "cache_missing_total", cacheMissingHelp, s.counterMetric, false},
```
Comment (joehandzik) [inline sources/procfs.go]: Probably better as 'cache_miss_total'

File: sources/procfs.go [hunk 5]
```diff
@@ -161,6 +180,24 @@ func (s *lustreProcfsSource) generateOSTMetricTemplates() {
...
+			{"encrypt_page_pools", "lowest_free_pages", lowFreeMarkHelp, s.gaugeMetric, false},
```
Comment (joehandzik) [inline sources/procfs.go]: free_page_low instead?

**Comment by joehandzik** [comment]
Suggestions cascade to all duplicated metrics, btw. Other than those tidbits, I think you're in good shape.

**Comment by roclark** [comment]
@joehandzik I realized what the issue was here with the tests failing. The path `/proc/fs/lustre/sptlrpc/encrypt_page_pools` doesn't uniquely identify a system or target, and this file can be located on any node, so the metrics were overrunning each other. I moved the metrics into "Generic" so that they can be parsed on every machine without troubles of overruning. Looks like everything is working as expected now.

**Comment by joehandzik** [comment]
@roclark Did we decide this one is ready to merge?

**Pull request merged**